File: apps/rhino-cli-go/internal/docs/fences.go
package docs

// CommonMark fenced-code-block tracking shared by the markdown gates.
//
// A fence opens on a run of >= 3 backticks or tildes (any leading
// whitespace tolerated) and closes ONLY on a same-character run of at least
// the opening length with nothing but whitespace after it (closing fences
// carry no info string; any leading whitespace tolerated). A naive toggle
// desyncs on nested constructs (a ````markdown block containing ```json
// inner fences) and on mixed fence characters (``` inside ~~~). Consumed by
// CollectATXHeadings (headings.go) and ExtractLinks (links_scanner.go).
// Mirrors the Rust `internal/docs/fences.rs` counterpart.
//
// Deliberate CommonMark deviation: the spec tolerates at most three leading
// spaces on a fence, but Prettier indents fences inside nested list items
// by 4-7 spaces — legitimate fences relative to the list container, which
// this flat (container-unaware) tracker would otherwise treat as text,
// causing false link/heading extractions from fence bodies (e.g.
// repo-governance/workflows/plan/plan-execution.md:524). We therefore
// accept ANY leading whitespace on opening AND closing fences. Trade-off:
// 4-space-indented code blocks whose content starts with ``` or ~~~ would
// be misread as fences — acceptable because Prettier never emits indented
// code blocks in this repository.

import "strings"

// fenceTracker tracks fenced-code-block state across the lines of one
// markdown document. The zero value is ready to use.
type fenceTracker struct {
	inFence bool
	char    byte // Opening fence character: '`' or '~'.
	length  int  // Opening fence run length (>= 3).
}

// observe consumes one line (without its trailing newline) and reports
// whether the consumer must skip it: fence delimiter lines (opening and
// closing) and every line inside an open fence return true.
func (f *fenceTracker) observe(line string) bool {
	char, length, rest, isFence := parseFenceRun(line)

	if f.inFence {
		// Close only on a same-char run of >= the opening length with no
		// info string (rest of line is whitespace).
		if isFence && char == f.char && length >= f.length && strings.TrimSpace(rest) == "" {
			f.inFence = false
		}
		return true
	}

	if isFence {
		f.inFence = true
		f.char = char
		f.length = length
		return true
	}

	return false
}

// parseFenceRun reports whether line begins (after any leading whitespace;
// see the deliberate CommonMark deviation in the package comment above)
// with a fence run of >= 3 backticks or tildes. On success it returns the
// fence character, the run length, and the remainder of the line after the
// run.
func parseFenceRun(line string) (char byte, length int, rest string, ok bool) {
	start := 0
	for start < len(line) && (line[start] == ' ' || line[start] == '\t') {
		start++
	}
	if start >= len(line) || (line[start] != '`' && line[start] != '~') {
		return 0, 0, "", false
	}

	char = line[start]
	end := start
	for end < len(line) && line[end] == char {
		end++
	}
	length = end - start
	if length < 3 {
		return 0, 0, "", false
	}

	return char, length, line[end:], true
}

File: apps/rhino-cli-rust/src/internal/docs/fences.rs
//! CommonMark fenced-code-block tracking shared by the heading and link
//! parsers. Mirrors the Go `internal/docs/fences.go` counterpart.
//!
//! Per the [CommonMark spec](https://spec.commonmark.org/0.31.2/#fenced-code-blocks)
//! a fence opened with N fence characters (backticks or tildes, N >= 3) is
//! closed ONLY by a fence of the SAME character with a run length >= N and no
//! info string. A naive "any ``` line toggles state" scan desyncs on nested
//! example fences (e.g. a ````markdown block containing ``` fences) and
//! misreads fence content as document content.
//!
//! **Deliberate deviation from strict CommonMark**: the spec caps fence
//! indentation at three leading spaces; this tracker accepts ANY leading
//! whitespace for both opening and closing fences. Prettier formats fences
//! inside nested list items at 4-7 space indents, which are legitimate
//! fences relative to their list container — a flat line tracker that
//! enforced the three-space cap would treat them as text and extract false
//! links/headings from fence content. The trade-off (a 4-space indented code
//! block would be mis-tracked as a fence) is acceptable because
//! Prettier-formatted markdown in this repo never produces indented code
//! blocks.

/// Stateful line-by-line fence tracker.
#[derive(Debug, Default)]
pub struct FenceTracker {
    /// Open fence, if any: `(fence_char, run_length)`.
    open: Option<(char, usize)>,
}

impl FenceTracker {
    /// Creates a tracker with no open fence.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Feeds the next line. Returns `true` when the line belongs to fenced
    /// code (an opening/closing delimiter line or interior fence content) and
    /// should therefore be excluded from markdown parsing.
    pub fn consume_line(&mut self, line: &str) -> bool {
        match self.open {
            Some((fence_char, open_len)) => {
                if is_closing_fence(line, fence_char, open_len) {
                    self.open = None;
                }
                // Delimiter and interior lines are both fence content.
                true
            }
            None => {
                if let Some(opened) = parse_opening_fence(line) {
                    self.open = Some(opened);
                    true
                } else {
                    false
                }
            }
        }
    }
}

/// Splits a fence-candidate line into `(fence_char, run_length, rest)` when
/// it starts with a backtick/tilde run after any leading whitespace (a
/// deliberate relaxation of CommonMark's three-space cap; see module docs).
fn split_fence_run(line: &str) -> Option<(char, usize, &str)> {
    let after_indent = line.trim_start();
    let fence_char = after_indent
        .chars()
        .next()
        .filter(|c| matches!(c, '`' | '~'))?;
    let run = after_indent
        .chars()
        .take_while(|&c| c == fence_char)
        .count();
    Some((fence_char, run, &after_indent[run..]))
}

/// Returns `Some((fence_char, run_length))` when `line` opens a fence: a run
/// of >=3 backticks or tildes after any leading whitespace. Backtick info
/// strings must not contain backticks (CommonMark); tilde info strings are
/// unrestricted.
fn parse_opening_fence(line: &str) -> Option<(char, usize)> {
    let (fence_char, run, rest) = split_fence_run(line)?;
    if run < 3 {
        return None;
    }
    if fence_char == '`' && rest.contains('`') {
        return None;
    }
    Some((fence_char, run))
}

/// Returns `true` when `line` closes a fence opened with `open_len` chars of
/// `fence_char`: same character, run length >= `open_len`, nothing but
/// whitespace after the run, any leading whitespace allowed.
fn is_closing_fence(line: &str, fence_char: char, open_len: usize) -> bool {
    let Some((run_char, run, rest)) = split_fence_run(line) else {
        return false;
    };
    run_char == fence_char && run >= open_len && rest.trim().is_empty()
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::panic)]
mod tests {
    use super::*;

    /// Runs the tracker over `content` and returns, per line, whether the
    /// line was classified as fence content.
    fn classify(content: &str) -> Vec<bool> {
        let mut tracker = FenceTracker::new();
        content
            .lines()
            .map(|line| tracker.consume_line(line))
            .collect()
    }

    #[test]
    fn simple_backtick_fence_opens_and_closes() {
        assert_eq!(
            classify("text\n```\ncode\n```\nafter\n"),
            vec![false, true, true, true, false]
        );
    }

    #[test]
    fn longer_open_requires_equal_or_longer_close() {
        // ``` cannot close ````; the final ```` does.
        assert_eq!(
            classify("````\n```\ncode\n````\nafter\n"),
            vec![true, true, true, true, false]
        );
    }

    #[test]
    fn closing_fence_must_match_char() {
        assert_eq!(
            classify("~~~\n```\n~~~\nafter\n"),
            vec![true, true, true, false]
        );
        assert_eq!(
            classify("```\n~~~\n```\nafter\n"),
            vec![true, true, true, false]
        );
    }

    #[test]
    fn closing_fence_rejects_info_string() {
        // A line with an info string never closes a fence.
        assert_eq!(
            classify("```\n``` rust\n```\nafter\n"),
            vec![true, true, true, false]
        );
    }

    #[test]
    fn deep_indented_fence_inside_list_item_is_tracked() {
        // Prettier indents fences inside nested list items 4-7 spaces; they
        // are legitimate fences relative to their list container. A 5-space
        // opener must suppress link/heading extraction until its 5-space
        // closer.
        assert_eq!(
            classify("- item\n     ```text\n     [not a link](./x.md)\n     ```\nafter\n"),
            vec![false, true, true, true, false]
        );
    }

    #[test]
    fn four_space_indent_is_treated_as_fence() {
        // Deliberate deviation from strict CommonMark: 4+ space indents open
        // fences here because Prettier-formatted markdown in this repo never
        // produces indented code blocks.
        assert_eq!(
            classify("    ```\ncode\n    ```\nafter\n"),
            vec![true, true, true, false]
        );
    }

    #[test]
    fn up_to_three_leading_spaces_allowed() {
        assert_eq!(
            classify("   ```\ncode\n   ```\nafter\n"),
            vec![true, true, true, false]
        );
    }

    #[test]
    fn backtick_info_string_with_backtick_is_not_a_fence() {
        assert_eq!(classify("``` a`b\ntext\n"), vec![false, false]);
    }

    #[test]
    fn two_chars_do_not_open_a_fence() {
        assert_eq!(classify("``\n~~\n"), vec![false, false]);
    }

    #[test]
    fn unclosed_fence_swallows_rest_of_document() {
        assert_eq!(classify("```\ncode\nmore\n"), vec![true, true, true]);
    }
}
